### ISSUE TITLE
Use api-key style

### ISSUE THREAD

**Issue by luixal** [kahmali/meteor-restivus#249]
Hi,

Let's say I have an API that I want some providers to use. Let's say I don't want those providers to have a username/password nor even need to log-in to use that API.

Is there any way to configure Restivus in an api-key style? This is, I generate a hash, I give that hash to one provider and that provider is able to use that API only including that api-key as a header.

Thanks :)

**Comment by kahmali** [comment]
>Is there any way to configure Restivus in an api-key style?

Yes, there is.

Sorry about the [lack of documentation for custom auth](https://github.com/kahmali/meteor-restivus/issues/98). The best explanation I've ever given can be found [here](https://github.com/kahmali/meteor-restivus/issues/129#issuecomment-133120800), although it goes into a lot of extra detail that needs to be stripped away. 

The 2nd config option [here](https://github.com/kahmali/meteor-restivus#configuration-options) is what you're looking for in the official docs (the README), but custom auth really deserves a dedicated section. I just [don't have enough time nowadays to work on Restivus](https://github.com/kahmali/meteor-restivus/issues/233#issuecomment-257134070), unfortunately.

**Comment by luixal** [comment]
I don't think we're talking about the same thing. I think about the ApiKey not as an auth system but a way to discard petition for not allowed apps.

It's just like, for example, twitter used to do: you used to register **your app** in twitter's dev site and you got an ApiKey **for your app** (the app itself) then someone could use your app and **auth himself with their credentials**.

Why I'm looking for this? Just to drop all queries without a valid ApiKey (i.e. DDOS attacks).

I guess I could do it with a custom auth function that check for an ApiKey in the headers before it tries to auth, but also seems like a think a feature to use _out of the box_.

Is there any function that gets called any time a request gets in? I could do a PR based on that.

**Comment by kahmali** [comment]
>Is there any function that gets called any time a request gets in?

YES! The [auth function](https://github.com/kahmali/meteor-restivus#configuration-options) I pointed out above is called for every endpoint where `{authRequired: true}` 😄 .

>I think about the ApiKey not as an auth system but a way to discard petition for not allowed apps.

"Discarding petition for not allowed apps" is essentially the definition of an auth system. You are ensuring that those that access your API are **auth**orized to do so.

>I don't think we're talking about the same thing.

We're absolutely talking about the same thing. You want to authorize someone based on something other than the default auth mechanism. The only way to do this is to override the default auth and validate the API key manually. I linked you to a description of how to configure auth for standard bearer tokens, which is essentially just a glorified API key. Just reread that [entire explanation I linked to above](https://github.com/kahmali/meteor-restivus/issues/129#issuecomment-133120800), but this time replace the words "standard bearer auth" with "API Key", and voila! 

Standard bearer tokens are generated differently from API keys, but that's just about where the differences stop. They're still both private tokens that are linked to a particular user and required to be passed with every API request to determine whether the request is allowed (authorized).

>I guess I could do it with a custom auth function that check for an ApiKey in the headers before it tries to auth, but also seems like a think a feature to use out of the box.

My goal (before I stopped having any time to work on Restivus) was to make it more modular so packages could be installed as add-ons with features like this. The majority of apps don't use a standard API key auth (some put the API key in the header or query params, some put it in the request body, they use different names for the key, or store the key in different locations in the database, etc). If anything, bearer token auth would be the one that came out-of-the-box, since it's such an accepted standard. If we keep adding every requested feature into the core package, it'll quickly become bloated, which I'd like to avoid. 

Hope that clears things up!
